src/s3/async_multipart_put_object.rs

```diff
@@ -157,6 +157,102 @@ impl<'a> AsyncMultipartUpload<'a> {
+    /// larger than 5MIB and smaller that 5GIB.
+    /// * `max_uploading_parts` - How many parts to upload concurrently,
+    /// Must be larger than 0 (defaults to 100).
+    #[instrument(skip(client))]
+    pub async fn from(
+        client: &'a Client,
+        upload_id: String,
+        dst: &S3Object,
+        part_size: usize,
+        max_uploading_parts: Option<usize>,
+    ) -> anyhow::Result<AsyncMultipartUpload<'a>> {
+        event!(Level::DEBUG, "New AsyncMultipartUpload");
+        if part_size < MIN_PART_SIZE {
+            anyhow::bail!("part_size was {part_size}, can not be less than {MIN_PART_SIZE}")
+        }
+        if part_size > MAX_PART_SIZE {
+            anyhow::bail!("part_size was {part_size}, can not be more than {MAX_PART_SIZE}")
+        }
+
+        //Check that user did not send in invalid parameter
+        if let Some(0) = max_uploading_parts {
+            anyhow::bail!("Max uploading parts must not be 0")
+        }
+
+        let mut parts: Vec<Part> = vec![];
+
+        let mut list_parts_result = client
+            .list_parts()
+            .bucket(&dst.bucket)
+            .key(&dst.key)
+            .upload_id(&upload_id)
+            .into_paginator()
+            .send();
+
+        // Use a paginator to collect all parts if there are more than 1000 parts in a multipart (the default and maximum part limit in list_parts())
+        while let Some(Ok(page)) = list_parts_result.next().await {
+            parts.append(&mut page.parts().unwrap_or_default().to_vec());
+
+            if !page.is_truncated() {
+                break;
+            }
+        }
+
+        let completed_parts: Vec<CompletedPart> = parts
+            .iter()
+            .map(|part| {
+                // Part has Option<&str> but CompletedPart has Option<String>
+                // Convert any &str's to String's but retain them as Options still
+
+                let e_tag = part.e_tag().map(|s| s.to_string());
+                let checksum_crc32 = part.checksum_crc32().map(|s| s.to_string());
+                let checksum_crc32_c = part.checksum_crc32_c().map(|s| s.to_string());
+                let checksum_sha1 = part.checksum_sha1().map(|s| s.to_string());
+                let checksum_sha256 = part.checksum_sha256().map(|s| s.to_string());
+
+                CompletedPart::builder()
+                    .set_e_tag(e_tag)
+                    .set_checksum_crc32(checksum_crc32)
+                    .set_checksum_crc32_c(checksum_crc32_c)
+                    .set_checksum_sha1(checksum_sha1)
+                    .set_checksum_sha256(checksum_sha256)
+                    .part_number(part.part_number())
+                    .build()
+            })
+            .collect();
+
+        let mut part_numbers: Vec<i32> = parts.iter().map(|p| p.part_number()).collect();
+        part_numbers.sort();
+
+        let latest_part_number = parts.iter().map(|p| p.part_number()).max().unwrap_or(0) + 1;
+
```

Comment:
Nice find with the `into_paginator()`.  I think this can be further reduced into this, which is using `items()` to give a flattened result of multiple calls to list items.

```suggestion
       let list_parts_result = client
            .list_parts()
            .bucket(&dst.bucket)
            .key(&dst.key)
            .upload_id(&upload_id)
            .into_paginator()
            .items()
            .send();

        let completed_parts: Vec<CompletedPart> = list_parts_result
            .map_ok(|part| {
                // Part has Option<&str> but CompletedPart has Option<String>
                // Convert any &str's to String's but retain them as Options still

                let e_tag = part.e_tag().map(|s| s.to_string());
                let checksum_crc32 = part.checksum_crc32().map(|s| s.to_string());
                let checksum_crc32_c = part.checksum_crc32_c().map(|s| s.to_string());
                let checksum_sha1 = part.checksum_sha1().map(|s| s.to_string());
                let checksum_sha256 = part.checksum_sha256().map(|s| s.to_string());

                CompletedPart::builder()
                    .set_e_tag(e_tag)
                    .set_checksum_crc32(checksum_crc32)
                    .set_checksum_crc32_c(checksum_crc32_c)
                    .set_checksum_sha1(checksum_sha1)
                    .set_checksum_sha256(checksum_sha256)
                    .part_number(part.part_number())
                    .build()
            })
            .try_collect()
            .await?;

        let latest_part_number = completed_parts
            .iter()
            .map(|p| p.part_number())
            .max()
            .unwrap_or(0)
            + 1;
```

Reply:
Also needs an import but I'm not sure how to do that in GitHub.
```Rust
use futures::TryStreamExt;
```